include/cuco/detail/hash_functions/identity_hash.cuh

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrust/functional.h>
+
+namespace cuco::detail {
+
+/**
+ * @brief An Identity hash function to hash the given argument on host and device
+ *
+ * @note `identity_hash` is perfect if `hash_table_capacity >= |input set|`
+ *
+ * @note `identity_hash` is only intended to be used perfectly.
+ *
+ * @note Perfect hashes are deterministic, and thus do not need seeds.
+ *
+ * @tparam Key The type of the values to hash
+ */
+template <typename Key>
+struct identity_hash : public thrust::identity<Key> {
+  using argument_type = Key;  ///< The type of the values taken as argument
+  using result_type   = Key;  ///< The type of the hash values produced
```

Comment:
One thing that just popped up in my head is that we assume a hash value to be of unsigned type for all of the other hashers. In our probing internals we need to perform some arithmetic combining (potentially signed) size types with unsigned hash value types. The logic as implemented has become quite complex, ensuring no signed integer overflow occurs, etc.

I'm only mildly concerned this is actually a problem since the unit test passes just fine, yet I wanted to bring this up just to make sure.

One other thing is that since we're using the hash value to do some integer arithmetics, we need to ensure the result type is actually an integer. Maybe just `reinterpret_cast<uint32_t or uint64_t>` the result of `thrust::identity` and call it a day?

Reply:
I think this points to a need for more user protection in `identity_hash` (and perfect hashing) in general. I believe we've discussed how to ensure `perfect_probing` always uses a perfect hash, but have we discussed whether perfect hashes can be used with other probing schemes? 
Here are my thoughts: 

# Option 1: `identity_hash` is ONLY intended for `perfect_probing`
It seems to me that `perfect_probing` would not require any arithmetic on the hash value for the following reasons:
- **No collisions occur.**
A hash value should indicate exactly one slot, and no others. No need to increment or otherwise alter the hash value to find a valid slot. 
- **Cooperative group logic is omitted.** 
Because we have no collisions, we shouldn't need cooperative groups either. We won't need to alter the hash value to enable cooperative probing. 
- **We already require `hash_table_size > max_possible_hash_value`, and `min_possible_hash_value >= 0`.** 
If this condition is not met, the hashing is not perfect. Presumably, we will have protections in place to ensure this requirement holds. 

If I'm correct about avoiding hash value arithmetic with `perfect_probing`, the only remaining way I can see `identity_hash` breaking in this mode is if the key type somehow cannot be converted into a valid slot. However, I would assume this means integer comparison won't work either, so it should fail the range test described above. 
In this mode, we would likely want protections to prevent users from using `identity_hash` with other probing schemes. 

# Option 2: We want `identity_hash` to work for non-perfect probing schemes
This mode seems closest to what a new user would assume, which may be a strong argument in its favor. In this mode, the points @sleeepyjack mentioned are extra important-- any output that isn't an integer will break the probing logic. But there are other problems as well: 
- **Double Hashing and perfect hashing don't cooperate** 
Any perfect hash that is supplied as both hashes to a `double_hashing` probing scheme will probably break (or at least perform terribly) on the first collision.  Because the prober should always return the same occupied slot, `insert` of the first collided value will either infinite loop or fail, depending on how that logic is implemented. Neither of these outcomes is acceptable-- therefore we likely need some protections to prevent users from using `double_hashing` with perfect hashers. 
- **Other probing can still be bad** 
Without the `is_perfect` guarantee, there are *zero* guarantees that `identity_hash` distributes values well. Note that every `key` that is a multiple of the `hash_table_size` hashes to zero due to the modulo operator. I imagine that most perfect hashes little-to-no performance guarantees for imperfect use.
- **Casting an `identity_hash` hash value could break perfect hashing** 
Certainly, if we allow types to pass through `identity_hash` that could break the prober, that's unacceptable. However, any of these problematic values would also break the `is_perfect` property if cast into to an acceptable range, because we now have two different keys that result in the same hash value. If go this route, I propose we limit the allowed `Key` types, or only cast if the probing scheme is not perfect. 


> I'm only mildly concerned this is actually a problem since the unit test passes just fine, yet I wanted to bring this up just to make sure.

I don't think my unit tests would catch this issue, as they only test the hasher and not any interaction with the probers. At the moment, the only protection in place is a comment stating `identity_hash` is only intended for use with `perfect probing`  🙂

Reply:
@sleeepyjack When I saw the modulo operator in the probing scheme without any `static_assert` that the output of the hash function is unsigned and integral, I knew there was a potential for breakage.

Just put two `static_assert` right next to that modulo operator and it should take care of itself, right?  You might also want to add unit tests.  Is it possible to make a unit test that passes when compilation fails?  You could test that using an identity hash that returns non-integral or signed values causes a compilation failure.

Reply:
After looking at the other hashers, it makes sense why `probing_scheme.impl` can assume an unsigned integral type: the existing hashers can only return `uint32_t` or `uint64_t`. If this responsibility belongs to the existing hashers, it makes sense for `identity_hash` to guarantee its output types as well. 

I was wrong to be worried about disrupting the `is_perfect` property due to signed types, since the bit pattern should remain the same. However, I think key types larger than `uint64_t` will still break the `is_perfect` property. Since `identity_hash` is more of an example perfect hash anyway, I think it makes sense to limit its input to sizes smaller than `uint64_t`.